### PR TITLE
GH-36621: [C++] Add documentation for ACERO_ALIGNMENT_HANDLING

### LINKED ISSUE #36621: [C++] Document ACERO_ALIGNMENT_HANDLING
### Describe the enhancement requested

This environment variable was added but we did not add documentation for it in the C++ user's guide.  We should document this variable so that users are aware how to configure it.  This is also a good chance to explain what this warning means in more detail.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by westonpace**
### Rationale for this change

To document an existing environment variable that was not documented.

### What changes are included in this PR?

Adds a documentation section for this environment variable.

### Are these changes tested?

No, there are only docs.

### Are there any user-facing changes?

New documentation
* Closes: #36621

### CODE REVIEW

**Comment by westonpace** [comment]
image

**Review by kou — APPROVED** [review]
+1

File: docs/source/cpp/env_vars.rst [hunk 1]
```diff
@@ -26,6 +26,28 @@ Arrow C++ at runtime.  Many of these variables are inspected only once per
...
+
+   Arrow C++'s Acero module performs computation on streams of data.  This
+   computation may involve a form of "type punning" that is technically
+   undefined behavior if the underlying array is not properly aligned.  On
+   most modern CPUs this is not an issue, but some older CPUs may crash or
+   suffer poor performance.  For this reason it is recommended that all
+   incoming array buffers are properly aligned.
+
+   The value of this environment variable controls what will happen when
+   Acero detects an unaligned buffer.
```
Comment (pitrou) [inline docs/source/cpp/env_vars.rst:39]:
```suggestion
   Acero detects an unaligned buffer:
```

File: docs/source/cpp/env_vars.rst [hunk 2]
```diff
@@ -26,6 +26,28 @@ Arrow C++ at runtime.  Many of these variables are inspected only once per
...
+
+   - ``warn`` a warning is emitted
+   - ``ignore`` nothing, alignment checking is disabled
+   - ``reallocate`` the buffer is reallocated to a properly aligned address
+   - ``error`` the operation will fail with an error
```
Comment (pitrou) [inline docs/source/cpp/env_vars.rst:44]:
```suggestion
   - ``warn``: a warning is emitted
   - ``ignore``: nothing, alignment checking is disabled
   - ``reallocate``: the buffer is reallocated to a properly aligned address
   - ``error``: the operation fails with an error
```

File: docs/source/cpp/env_vars.rst [hunk 3]
```diff
@@ -26,6 +26,28 @@ Arrow C++ at runtime.  Many of these variables are inspected only once per
...
 that changing their value later will have an effect.
 
+.. envvar:: ACERO_ALIGNMENT_HANDLING
+
+   Arrow C++'s Acero module performs computation on streams of data.  This
+   computation may involve a form of "type punning" that is technically
+   undefined behavior if the underlying array is not properly aligned.  On
+   most modern CPUs this is not an issue, but some older CPUs may crash or
+   suffer poor performance.  For this reason it is recommended that all
+   incoming array buffers are properly aligned.
```
Comment (pitrou) [inline docs/source/cpp/env_vars.rst:36]:
```suggestion
   suffer poor performance.  For this reason it is recommended that all
   incoming array buffers are properly aligned, but some data sources
   such as :ref:`Flight <flight-rpc>` may produce unaligned buffers.
```

**Review by pitrou — APPROVED** [review]
Thanks for this! Some small suggestions for improvement below.

**Comment by kou** [comment]
@github-actions crossbow submit preview-docs

**Comment by kou** [comment]
http://crossbow.voltrondata.com/pr_docs/36622/cpp/env_vars.html

I'll merge this.

**Comment by RichardHaythorn** [comment]
@westonpace I see it's already merged but it LGTM

**Pull request merged**